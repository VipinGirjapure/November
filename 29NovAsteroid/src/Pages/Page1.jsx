import {
  Backdrop,
  Button,
  CircularProgress,
  Grid,
  TextField,
} from "@mui/material";
import React, { useState } from "react";
import { useNavigate } from "react-router-dom";

function Page1() {
  const [loading, setLoading] = useState(false);
  const [astroid, setAstroid] = useState("");
  let navigate = useNavigate();

  const apiKey = "0cDupsui0V2X1B412ivprzPhKjMAHazLgzlsVICV";

  const astroidApi = (id) => {
    const url = `https://api.nasa.gov/neo/rest/v1/neo/${id}?api_key=${apiKey}`;
    setLoading(true);

    fetch(url)
      .then((response) => response.json())
      .then((result) => {
        // console.log("result", result);
        navigate("/page2", {
          state: { result },
        });
      })
      .catch((e) => console.log("Please Enter Valid Astroid", e));
  };

  const randomAstroid = () => {
    setLoading(true);
    const urlkey = `https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=${apiKey}`;
    fetch(urlkey)
      .then((resp) => resp.json())
      .then((res) => {
        let item =
          res.near_earth_objects[
            Math.floor(Math.random() * res.near_earth_objects.length)
          ];

        astroidApi(item.id);
      })
      .catch((e) => console.log("error", e));
  };

  return (
    <Grid align="center" sx={{ marginTop: "15%" }}>
      <TextField
        data-testid="asteroidId"
        placeholder="Enter Astroid Id"
        value={astroid}
        onChange={(e) => setAstroid(e.target.value)}
      />
      <br />

      <Button
        data-testid="submitBtn"
        sx={{ margin: "2%" }}
        variant="contained"
        onClick={() => astroidApi(astroid)}
        disabled={astroid.length <= 1}
      >
        Submit
      </Button>
      <br />

      <Button
        variant="contained"
        onClick={randomAstroid}
        data-testid="randomId"
      >
        Random Astroid
      </Button>

      <Backdrop
        sx={{ color: "#fff", zIndex: (theme) => theme.zIndex.drawer + 1 }}
        open={loading}
      >
        {loading && <CircularProgress color="inherit" data-testid="loader" />}
      </Backdrop>
    </Grid>
  );
}

export default Page1;
