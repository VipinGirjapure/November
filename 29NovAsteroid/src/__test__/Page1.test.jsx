import { fireEvent, render, screen, act } from "@testing-library/react";
import Page1 from "../Pages/Page1";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => {
  return {
    useNavigate: () => mockNavigate,
  };
});

global.alert = (e) => {
  console.log("error", e);
};

describe("Page1 test suit", () => {
  beforeEach(() => {
    jest.spyOn(window, "fetch").mockImplementationOnce((url) => {
      if (
        url ===
        "https://api.nasa.gov/neo/rest/v1/neo/123?api_key=0cDupsui0V2X1B412ivprzPhKjMAHazLgzlsVICV"
      ) {
        return Promise.reject("please enter valid asteroid id");
      }
      return Promise.resolve({
        json: () =>
          Promise.resolve({
            id: 123,
          }),
      });
    });
  });

  test("render Page1 component correctly", () => {
    render(<Page1 />);
    expect(screen).toBeTruthy();
  });

  test("Should value change when user enter any value in asteroid id text field", () => {
    const { getByTestId } = render(<Page1 />);
    const asteroidId = getByTestId("asteroidId").querySelector("input");

    fireEvent.change(asteroidId, { target: { value: "12345" } });
    expect(asteroidId.value).toBe("12345");
  });

  test("try to get asteroid data when user enter wrong id", () => {
    const { getByTestId } = render(<Page1 />);
    const submitBtn = getByTestId("submitBtn");
    const asteroidId = getByTestId("asteroidId").querySelector("input");

    expect(submitBtn).toBeDisabled();

    fireEvent.change(asteroidId, { target: { value: "123" } });

    expect(submitBtn).toBeEnabled();

    fireEvent.click(submitBtn);

    const loader = getByTestId("loader");
    expect(loader).toBeDefined();
  });

  test("try to get asteroid data when user enter correct id", async () => {
    const { getByTestId } = render(<Page1 />);
    const submitBtn = getByTestId("submitBtn");
    const asteroidId = getByTestId("asteroidId").querySelector("input");
    await act(async () => {
      await fireEvent.change(asteroidId, { target: { value: "123567" } });
      await fireEvent.click(submitBtn);
    });

    expect(mockNavigate).toBeCalledWith("/page2", {
      state: { result: expect.any(Object) },
    });
  });
});

test("random asteroid id", async () => {
  const { getByTestId } = render(<Page1 />);
  const randomId = getByTestId("randomId");

  await act(async () => {
    fireEvent.click(randomId);
  });
});

global.alert = jest.fn();

const mockData = [
  {
    name: "n1",
    email: "test@maul.com",
  },
  {
    name: "n2",
    email: "test2@maul.com",
  },
];

global.fetch = jest.fn().mockImplementationOnce(() => {
  return new Promise((resolve, reject) => {
    reject("network error");
  });
});

describe("App.tsx test suit", () => {
  beforeEach(() => {
    global.fetch = jest
      .fn()

      .mockImplementation(() => {
        return new Promise((resolve, reject) => {
          resolve({
            json: () =>
              new Promise((res, rej) => {
                res(mockData);
              }),
          });
        });
      });
  });

  it("snapshot testing", () => {
    const { container } = render(<Page1 />);
    expect(container).toMatchSnapshot();
  });

  it("Should error handle when got error from fetch", async () => {
    const { getByTestId } = render(<Page1 />);
    const randomId = getByTestId("randomId");
    await act(async () => {
      await fireEvent.click(randomId);
    });
    // const row = getByTestId('row-0')
    // expect(row).toBeDefined()
  });
});
