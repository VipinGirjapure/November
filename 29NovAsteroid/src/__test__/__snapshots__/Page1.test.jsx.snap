// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.tsx test suit snapshot testing 1`] = `
<div>
  <div
    align="center"
    class="MuiGrid-root css-bs0e0n-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="asteroidId"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r5:"
          placeholder="Enter Astroid Id"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <br />
    <button
      class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-150vanc-MuiButtonBase-root-MuiButton-root"
      data-testid="submitBtn"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Submit
    </button>
    <br />
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="randomId"
      tabindex="0"
      type="button"
    >
      Random Astroid
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;
